bot/cogs/doc.py

```diff
@@ -0,0 +1,546 @@
+import asyncio
+import functools
+import logging
+import random
+import re
+from collections import OrderedDict
+from ssl import CertificateError
+from typing import Dict, List, Optional, Tuple
+
+import discord
+from aiohttp import ClientConnectorError
+from bs4 import BeautifulSoup
+from discord.ext import commands
+from markdownify import MarkdownConverter
+from requests import ConnectionError
+from sphinx.ext import intersphinx
+
+from bot.constants import (
+    ADMIN_ROLE, ERROR_REPLIES, MODERATOR_ROLE,
+    OWNER_ROLE, SITE_API_KEY, SITE_API_URL
+)
+from bot.decorators import with_role
+from bot.exceptions import CogBadArgument
+
+
+log = logging.getLogger(__name__)
+logging.getLogger('urllib3').setLevel(logging.WARNING)
+
+
+UNWANTED_SIGNATURE_SYMBOLS = ('[source]', '¶')
+WHITESPACE_AFTER_NEWLINES_RE = re.compile(r"(?<=\n\n)( +)")
+
+
+def async_cache(max_size=128, arg_offset=0):
+    """
+    LRU cache implementation for coroutines.
+
+    :param max_size:
+    Specifies the maximum size the cache should have.
+    Once it exceeds the maximum size, keys are deleted in FIFO order.
+    :param arg_offset:
+    The offset that should be applied to the coroutine's arguments
+    when creating the cache key. Defaults to `0`.
+    """
+
+    # Assign the cache to the function itself so we can clear it from outside.
+    async_cache.cache = OrderedDict()
+
+    def decorator(function):
+        @functools.wraps(function)
+        async def wrapper(*args):
+            key = ':'.join(args[arg_offset:])
+
+            value = async_cache.cache.get(key)
+            if value is None:
+                if len(async_cache.cache) > max_size:
+                    async_cache.cache.popitem(last=False)
+
+                async_cache.cache[key] = await function(*args)
+            return async_cache.cache[key]
+        return wrapper
+    return decorator
+
+
+class DocMarkdownConverter(MarkdownConverter):
+    def convert_code(self, el, text):
+        """Undo `markdownify`s underscore escaping."""
+
+        return f"`{text}`".replace('\\', '')
+
+    def convert_pre(self, el, text):
+        """Wrap any codeblocks in `py` for syntax highlighting."""
+
+        code = ''.join(el.strings)
+        return f"```py\n{code}```"
+
+
+def markdownify(html):
+    return DocMarkdownConverter(bullets='•').convert(html)
+
+
+class DummyObject(object):
+    """
+    A dummy object which supports assigning anything,
+    which the builtin `object()` does not support normally.
+    """
+
+
+class SphinxConfiguration:
+    """Dummy configuration for use with intersphinx."""
+
+    config = DummyObject()
+    config.intersphinx_timeout = 3
+    config.tls_verify = True
+
+
+class ValidPythonIdentifier(commands.Converter):
```

Comment:
Same with these converters. Though they're only used here, it would be cleaner to have them in their own file with this file only containing the cog.